include/seqan3/core/range/type_traits.hpp

```diff
@@ -0,0 +1,287 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides various transformation traits used by the range module.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+#include <type_traits>
+
+#include <seqan3/core/detail/iterator_traits.hpp>
+#include <seqan3/core/platform.hpp>
+#include <seqan3/utility/type_traits/basic.hpp>
+#include <seqan3/utility/type_traits/pre.hpp>
+
+// TODO(h-2): add innermost_reference instead of or addition to range_innermost_value?
+
+//NOTE(h-2): for the range overloads we explicitly forbid that the type is iteratoer
+// because some types are actually both (e.g. std::directory_iterator)
+
+namespace seqan3::detail
+{
+
+//!\cond
+template <typename t>
+SEQAN3_CONCEPT has_range_value_type = requires { typename std::ranges::range_value_t<std::remove_cvref_t<t>>; };
+//!\endcond
+
+//!\brief Makes range_t const if const_range is true; otherwise keeps range_t as is.
+//!\ingroup type_traits
+template <bool const_range, typename range_t>
+using maybe_const_range_t = std::conditional_t<const_range, range_t const, range_t>;
+
+//!\brief Returns the const iterator of range_t if const_range is true; otherwise the non-const iterator.
+//!\ingroup type_traits
+template <bool const_range, typename range_t>
+using maybe_const_iterator_t = std::ranges::iterator_t<maybe_const_range_t<const_range, range_t>>;
+
+//!\brief Returns the const sentinel of range_t if const_range is true; otherwise the non-const sentinel.
+//!\ingroup type_traits
+template <bool const_v, typename range_t>
+using maybe_const_sentinel_t = std::ranges::sentinel_t<maybe_const_range_t<const_v, range_t>>;
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+/*!\addtogroup type_traits
+ * \{
+ */
+
+// ----------------------------------------------------------------------------
+// value_type
+// ----------------------------------------------------------------------------
+
+#ifdef SEQAN3_DEPRECATED_310
```

Comment:
why did you deprecate all of this? I thought only range_compatible should be deprecated?

This would have been easier if it would have been done in a separate PR ;) for next time.

Reply:
These `#ifdef SEQAN3_DEPRECATED_310` were already in the `type_traits/range.hpp` before and I didn't decide to deprecate this. I simply moved the whole file without changing its content (except removing range_compatible in the following commit). If you look at the individual commits it should become clearer where I moved a file or where I changed content.

Reply:
Ah ok! sorry I missed that :)